src/java/picard/sam/revertsam/RevertSam.java

```diff
@@ -148,75 +144,51 @@ public static void main(final String[] args) {
-                        "A value for SAMPLE_ALIAS cannot be supplied.");
-            }
-            if (LIBRARY_NAME != null && !allLibraryNamesIdentical) {
-                throw new PicardException("Read groups have multiple values for library name.  " +
-                        "A value for library name cannot be supplied.");
-            }
+        OutputMap outputMap = null;
+        if (OUTPUT_BY_READGROUP) {
+            final String defaultExtension = getDefaultExtension(INPUT.toString());
+            outputMap = RevertSamUtil.createOutputMap(OUTPUT_MAP, OUTPUT, defaultExtension, inHeader);
+            ValidationUtil.assertAllReadGroupsMapped(outputMap, inHeader.getReadGroups());
         }
 
         ////////////////////////////////////////////////////////////////////////////
         // Build the output writer with an appropriate header based on the options
         ////////////////////////////////////////////////////////////////////////////
-        final boolean presorted = (inHeader.getSortOrder() == SORT_ORDER) || (SORT_ORDER == SortOrder.queryname && SANITIZE);
-        final SAMFileHeader outHeader = new SAMFileHeader();
-        for (final SAMReadGroupRecord rg : inHeader.getReadGroups()) {
-            if (SAMPLE_ALIAS != null) {
-                rg.setSample(SAMPLE_ALIAS);
-            }
-            if (LIBRARY_NAME != null) {
-                rg.setLibrary(LIBRARY_NAME);
-            }
-            outHeader.addReadGroup(rg);
-        }
-        outHeader.setSortOrder(SORT_ORDER);
-        if (!REMOVE_ALIGNMENT_INFORMATION) {
-            outHeader.setSequenceDictionary(inHeader.getSequenceDictionary());
-            outHeader.setProgramRecords(inHeader.getProgramRecords());
-        }
-
-        final SAMFileWriter out = new SAMFileWriterFactory().makeSAMOrBAMWriter(outHeader, presorted, OUTPUT);
+        final boolean presorted = isPresorted(inHeader, SORT_ORDER, sanitizing);
+        List<ReadGroup> readGroups = null;
+        if (OUTPUT_BY_READGROUP) {
+            readGroups = RevertSamUtil.makeReadGroupList(inHeader, outputMap, SORT_ORDER, REMOVE_ALIGNMENT_INFORMATION);
+        } 
+        final SAMFileHeader singleOutHeader = RevertSamUtil.createSingleOutHeader(SORT_ORDER, inHeader, REMOVE_ALIGNMENT_INFORMATION);
+        if (SAMPLE_ALIAS != null) overrideSample(singleOutHeader, SAMPLE_ALIAS);
+        if (LIBRARY_NAME != null) overrideLibrary(singleOutHeader, LIBRARY_NAME);
+        final RevertSamWriter out = RevertSamUtil.makeSingleOrPerReadGroupWriter(inHeader, singleOutHeader, presorted, OUTPUT_BY_READGROUP, readGroups, OUTPUT);
 
         ////////////////////////////////////////////////////////////////////////////
         // Build a sorting collection to use if we are sanitizing
         ////////////////////////////////////////////////////////////////////////////
-        final SortingCollection<SAMRecord> sorter;
+        RevertSamSorter sorter = null;
```

Comment:
Instead of a whole new set of classes, why not just two maps:
```
final Map<String, PeekableIterator<SAMRecord> iterators;
final Map<String, SortingCollection<SAMRecord> sorters;
```
In the case of one output, just have them map to the same iterator or sorter (you'll need to unique them later), or just condition on `OUTPUT_BY_READGROUP` laster.  There are probably other ways, but are three new classes to store two simple maps necessary?